File: apps/mobile/test/widgets/star_rating_test.dart
import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';
import 'package:mobile/core/api_client.dart';
import 'package:mobile/models/concert.dart';
import 'package:mobile/widgets/star_rating.dart';

import '../support/fake_api_client.dart';

void main() {
  testWidgets('taper sur une étoile envoie la note et déclenche onRated', (
    tester,
  ) async {
    String? ratedConcertId;
    int? ratedValue;
    var onRatedCalled = false;
    final api = FakeApiClient()
      ..onRateConcert = (id, value) async {
        ratedConcertId = id;
        ratedValue = value;
      };

    await tester.pumpWidget(
      MaterialApp(
        home: Scaffold(
          body: StarRating(
            api: api,
            concertId: 'c1',
            summary: const ConcertRatingSummary(average: null, count: 0),
            onRated: () => onRatedCalled = true,
          ),
        ),
      ),
    );

    await tester.tap(find.byTooltip('4 étoiles'));
    await tester.pumpAndSettle();

    expect(ratedConcertId, 'c1');
    expect(ratedValue, 4);
    expect(onRatedCalled, isTrue);
    expect(find.text('Merci, votre note a été enregistrée.'), findsOneWidget);
  });

  testWidgets('une erreur API affiche le message au lieu du succès', (
    tester,
  ) async {
    final api = FakeApiClient()
      ..onRateConcert = (id, value) async =>
          throw const ApiException(500, 'Erreur serveur');

    await tester.pumpWidget(
      MaterialApp(
        home: Scaffold(
          body: StarRating(
            api: api,
            concertId: 'c1',
            summary: const ConcertRatingSummary(average: null, count: 0),
            onRated: () {},
          ),
        ),
      ),
    );

    await tester.tap(find.byTooltip('3 étoiles'));
    await tester.pumpAndSettle();

    expect(find.text('Erreur serveur'), findsOneWidget);
    expect(find.text('Merci, votre note a été enregistrée.'), findsNothing);
  });
}

File: apps/web/src/lib/components/concert/StarRating.test.ts
import { describe, expect, it, vi } from 'vitest';
import { render, screen, fireEvent, waitFor } from '@testing-library/svelte';
import StarRating from './StarRating.svelte';
import { ApiError } from '$lib/api/client';

const { rateConcert } = vi.hoisted(() => ({ rateConcert: vi.fn() }));

vi.mock('$lib/api/client', async (importOriginal) => {
	const actual = await importOriginal<typeof import('$lib/api/client')>();
	return { ...actual, api: { ...actual.api, rateConcert } };
});

describe('StarRating', () => {
	it('cliquer sur une étoile envoie la note', async () => {
		rateConcert.mockResolvedValue(undefined);
		render(StarRating, { concertId: 'c1', summary: { average: null, count: 0 } });

		await fireEvent.click(screen.getByRole('radio', { name: '4 étoiles' }));

		await waitFor(() => expect(rateConcert).toHaveBeenCalledWith('c1', 4));
		expect(screen.getByText('Merci, votre note a été enregistrée.')).toBeInTheDocument();
	});

	it('une erreur API affiche le message au lieu du succès', async () => {
		rateConcert.mockRejectedValue(new ApiError(500, 'Erreur serveur'));
		render(StarRating, { concertId: 'c1', summary: { average: null, count: 0 } });

		await fireEvent.click(screen.getByRole('radio', { name: '3 étoiles' }));

		await waitFor(() => expect(screen.getByText('Erreur serveur')).toBeInTheDocument());
		expect(screen.queryByText('Merci, votre note a été enregistrée.')).not.toBeInTheDocument();
	});
});
